File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        ans[n - 1] = 1;
        for(int i = n - 2; i >= 0; i--) {
            ans[i] = ans[i + 1] * nums[i + 1];
        }
        int curr = 1;
        for(int i = 0; i < n; i++) {
            ans[i] *= curr;
            curr *= nums[i];
        }
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        prevProd = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] = prevProd 
            prevProd = prevProd * nums[i]
        
        prevProd = 1
        for i, num in enumerate(nums):
            res[i] = prevProd * res[i]
            prevProd = prevProd * num
        return res
'''
nums = [1, 2, 3, 4]

brute force approach
  o(n x n) o(1)

precalucate suffix product
[ 1, 2,3,4]
[12,12,4,1]

prefixSum = 1

'''